docs/modules/ROOT/pages/design/database-interface.adoc

```diff
@@ -0,0 +1,105 @@
+= Database Interface
+
+== Overview
+
+This page documents the proposal for the structure and content of `pkg/database` to achieve the following goals:
+
+* Reduce the friction when adding support to a new DBMS(Database Management System).
+* Enable overriding the default read/write operations for a specific DBMS.
+* Expose the filters (`WHERE` clauses) and sorters (`ORDER BY` clauses) separately from the rest of the `SQL` statements
+so they can be plugged in and out of the different queries.
+* Reduce the maintenance of the codebase.
+* Increase the readability of the codebase.
+
+== Proposal
+
+The two pillars of this proposal are:
+
+* Definition of interfaces that need to be implementing when adding support to a new DBMS.
+* Usage of a SQL Builder.
+
+=== Database Interfaces
+
+To add a new DBMS the following interfaces must be implemented:
+
+`DBInterface`::
+Include the methods that should be used by the rest of the KubeArchive application.
+Encapsulates the business logic related to the different read/write database operations:
+`QueryResources`, `WriteResources`, `WriteUrls`, `QueryLogUrls`, etc.
+
+`DBCreator`::
+Include the methods `GetDriverName()` and `GetConnectionString()`.
+They are used for establishing the database connection in the `DBInterface` constructor.
+
+`DBSelector`::
+Encapsulate the different `SELECT` SQL statements without the `WHERE` and `ORDER BY` clauses.
+
+`DBFilter`::
+Encapsulate the different `WHERE` SQL clauses that can be used in `SELECT`, `DELETE` or `UPDATE` statements.
+
+`DBSorter`::
+Encapsulate the different `ORDER BY` clauses.
+
+`DBInserter`::
+Encapsulate the different `INSERT` SQL statements.
+
+`DBDeleter`::
+Encapsulate the different `DELETE` SQL statement without the `WHERE` clauses.
```

Comment:
Can or should we force `DBDeleter` to have a `WHERE` clause?  It seems dangerous not to.

We might also want to do it for `DBSelector`, but that's not quite as dangerous as `DBDeleter`.

Reply:
So, the idea is that the SQL statements are separated so the `WHERE` clauses can be plugged in when needed.
Then, the same `WHERE` clauses can be used in a `DBSelector` and in a `DBDeleter` (and `DBUpdater` if we have one in the future). So the idea I had here was the following: using a sqlbuilder doesn't restrict you for writing any type of query, and the responsability of writing good queries rely in the `database.go` functions that specify the different parts that are used in the query.

From my POV, we should have tests for checking that we don't mess things up.

Reply:
My thinking was just having the `DBDeleter` requiring a `WHERE` clause be passed to it, so you could still use the same `WHERE` in multiple places if you like.

But I'm OK with not changing things.

Reply:
I see... the problem I see with that is that considering `go-sqlbuilder` architecture to create a `WHERE` clause  you need a sqlbuilder (i.e, a `DBSelector` a `DBDeleter` and a `DBUpdater`) but not the other way around... so I don't know how to accomplish your suggestion without doing complicated stuff

Reply:
Yeah, leave it. I've already approved.